src/py/flwr/server/strategy/fedavginplace.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2023 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""FedAvg with in-place aggregation."""
+
+
+from functools import reduce
+from logging import WARNING
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+
+from flwr.common import (
+    FitRes,
+    Parameters,
+    Scalar,
+    ndarrays_to_parameters,
+    parameters_to_ndarrays,
+)
+from flwr.common.logger import log
+from flwr.server.client_proxy import ClientProxy
+
+from .fedavg import FedAvg
+
+
+class FedAvgInPlace(FedAvg):
+    """FedAvg with in-place aggregation."""
+
+    def __repr__(self) -> str:
+        """Compute a string representation of the strategy."""
+        rep = f"FedAvgInPlace(accept_failures={self.accept_failures})"
+        return rep
+
+    def aggregate_fit(
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, FitRes]],
+        failures: List[Union[Tuple[ClientProxy, FitRes], BaseException]],
+    ) -> Tuple[Optional[Parameters], Dict[str, Scalar]]:
+        """Aggregate fit results using weighted average."""
+        if not results:
+            return None, {}
+        # Do not aggregate if there are failures and failures are not accepted
+        if not self.accept_failures and failures:
+            return None, {}
+
+        # Count total examples
+        num_examples_total = sum([fit_res.num_examples for _, fit_res in results])
+
+        # Compute scaling factors for each result
+        scaling_factors = [
+            fit_res.num_examples / num_examples_total for _, fit_res in results
+        ]
+
+        # Let's do in-place aggregation
+        # get first result, then add up each other
+        params = [
+            scaling_factors[0] * x
+            for x in parameters_to_ndarrays(results[0][1].parameters)
+        ]
+        for i, (_, fit_res) in enumerate(results[1:]):
+            res = (
+                scaling_factors[i + 1] * x
+                for x in parameters_to_ndarrays(fit_res.parameters)
+            )
+            params = [
+                reduce(np.add, layer_updates) for layer_updates in zip(params, res)
+            ]
```

Comment:
Should we move this to an `aggregate_inplace(results: List[Tuple[ClientProxy, FitRes]])` function sitting in `aggregate.py` ?

Reply:
Yeah. But how we make it easy to use? adding a new `inplace` flag to the existing `FedAvg` that would call this `aggregate_inplace` instead of doing what currently `FedAvg` does in the first half of its `aggregate_fit()`?

Reply:
I personally like the idea of the `inplace` flag on the existing `FedAvg` that would call the `aggregate_inplace` function. 
One benefit would also be that we make the aggregation step a bit more explicit. I noticed quite a few people that where going through `FedAvg` and struggled to figure out that the aggregation happens in `aggregate.py`. I attribute it to the fact that `aggregate` is called inside `ndarrays_to_parameters` which doesn't jump to the eye. Whereas if we have a check for the aggregation method, it will be more visible.

Reply:
I made the change. I also moved the post-aggregation conversion `ndarrays_to_parameters` in the existing implementation.